### PR TITLE
Update to ruma-api v0.3.

### LINKED ISSUE #14: Upgrade to ruma-api 0.3.0
[ruma-api 0.3.0](https://github.com/ruma/ruma-api/releases/tag/0.3.0) includes a type signature change for `Endpoint::router_path` and adds four new required methods. We'll need to update all the endpoints here.

### PR DESCRIPTION
**Pull request by vberger**
Also fixes a few errors encountered in the process.

I took the liberty to rename the endpoints related to state events into something more explicit about their function. This is not final and probably requires bikeshedding, as per #10 

fixes #14

### CODE REVIEW

File: src/r0/account.rs [hunk 1]
```diff
@@ -108,18 +108,36 @@ pub mod register {
...
+        fn router_path() -> &'static str {
+            "/_matrix/client/r0/register"
+        }
+
+        fn name() -> &'static str {
+            "register"
+        }
+
+        fn description() -> &'static str {
+            "Register an account to this homeserver."
```
Comment (jimmycuadra) [inline src/r0/account.rs]: to --> on

File: src/r0/account.rs [hunk 2]
```diff
@@ -145,11 +163,27 @@ pub mod request_password_change_token {
...
         }
 
-        fn router_path() -> String {
-            "/_matrix/client/r0/account/password/email/requestToken".to_string()
+        fn name() -> &'static str {
+            "request_password_change_token"
+        }
+
+        fn description() -> &'static str {
+            "Request that a password change token is sent to given email address."
```
Comment (jimmycuadra) [inline src/r0/account.rs]: the given email address

File: src/r0/account.rs [hunk 3]
```diff
@@ -173,11 +207,27 @@ pub mod deactivate {
...
+        }
+
+        fn name() -> &'static str {
+            "deactivate"
         }
 
-        fn router_path() -> String {
-            "/_matrix/client/r0/account/deactivate".to_string()
+        fn description() -> &'static str {
+            "Deactivate the account."
```
Comment (jimmycuadra) [inline src/r0/account.rs]: the --> an

File: src/r0/account.rs [hunk 4]
```diff
@@ -206,11 +256,27 @@ pub mod change_password {
...
+        fn router_path() -> &'static str {
+            "/_matrix/client/r0/account/password"
+        }
+
+        fn name() -> &'static str {
+            "change_password"
+        }
+
+        fn description() -> &'static str {
+            "Change the password of the account."
```
Comment (jimmycuadra) [inline src/r0/account.rs]: an account

File: src/r0/config.rs [hunk 5]
```diff
@@ -74,8 +90,24 @@ pub mod set_global_account_data  {
...
+        fn router_path() -> &'static str {
+            "/_matrix/client/r0/user/:user_id/account_data/:type"
+        }
+
+        fn name() -> &'static str {
+            "set_global_account_data"
+        }
+
+        fn description() -> &'static str {
+            "Sets the global account data."
```
Comment (jimmycuadra) [inline src/r0/config.rs]: Sets the global --> Set global

File: src/r0/context.rs [hunk 6]
```diff
@@ -48,8 +48,24 @@ pub mod get_context {
...
+        fn router_path() -> &'static str {
+            "/_matrix/client/r0/rooms/:room_id/context/:event_id"
+        }
+
+        fn name() -> &'static str {
+            "get_context"
+        }
+
+        fn description() -> &'static str {
+            "Get the events immediatly preceding and following a given event."
```
Comment (jimmycuadra) [inline src/r0/context.rs]: immediately

File: src/r0/directory.rs [hunk 7]
```diff
@@ -43,11 +43,27 @@ pub mod get_public_rooms {
...
+        fn router_path() -> &'static str {
+            "/_matrix/client/r0/publicRooms"
+        }
+
+        fn name() -> &'static str {
+            "get_public_rooms"
+        }
+
+        fn description() -> &'static str {
+            "Get the list of rooms in this homeserver public directory."
```
Comment (jimmycuadra) [inline src/r0/directory.rs]: homeserver's

File: src/r0/media.rs [hunk 8]
```diff
@@ -34,8 +34,24 @@ pub mod get_content {
...
+        fn router_path() -> &'static str {
+            "/_matrix/media/r0/download/:server_name/:media_id"
+        }
+
+        fn name() -> &'static str {
+            "get_media_content"
+        }
+
+        fn description() -> &'static str {
+            "Retrieve a content from the media store."
```
Comment (jimmycuadra) [inline src/r0/media.rs]: Retrieve content

File: src/r0/media.rs [hunk 9]
```diff
@@ -64,11 +80,28 @@ pub mod create_content {
...
+        }
+
+        fn name() -> &'static str {
+            "create_media_content"
         }
 
-        fn router_path() -> String {
-            "/_matrix/media/r0/upload".to_string()
+        fn description() -> &'static str {
+            "Upload a content to the media store."
```
Comment (jimmycuadra) [inline src/r0/media.rs]: Upload content

File: src/r0/membership.rs [hunk 10]
```diff
@@ -55,8 +55,24 @@ pub mod invite_user {
...
+        fn router_path() -> &'static str {
+            "/_matrix/client/r0/rooms/:room_id/invite"
+        }
+
+        fn name() -> &'static str {
+            "invite_user"
+        }
+
+        fn description() -> &'static str {
+            "Invite an user to a room."
```
Comment (jimmycuadra) [inline src/r0/membership.rs]: a user

File: src/r0/membership.rs [hunk 11]
```diff
@@ -116,8 +132,24 @@ pub mod join_room_by_id_or_alias {
...
+        fn router_path() -> &'static str {
+            "/_matrix/client/r0/join/:room_id_or_alias"
+        }
+
+        fn name() -> &'static str {
+            "join_room_by_id_or_alias"
+        }
+
+        fn description() -> &'static str {
+            "Join a room, using it's ID or one of its aliases."
```
Comment (jimmycuadra) [inline src/r0/membership.rs]: remove comma

it's --> its

File: src/r0/membership.rs [hunk 12]
```diff
@@ -167,8 +199,24 @@ pub mod join_room_by_id {
...
+        fn router_path() -> &'static str {
+            "/_matrix/client/r0/rooms/:room_id/join"
+        }
+
+        fn name() -> &'static str {
+            "join_room_by_id"
+        }
+
+        fn description() -> &'static str {
+            "Join a room using it's ID."
```
Comment (jimmycuadra) [inline src/r0/membership.rs]: it's --> its

File: src/r0/membership.rs [hunk 13]
```diff
@@ -286,8 +366,24 @@ pub mod kick_user {
...
+        fn router_path() -> &'static str {
+            "/_matrix/client/r0/rooms/:room_id/kick"
+        }
+
+        fn name() -> &'static str {
+            "kick_user"
+        }
+
+        fn description() -> &'static str {
+            "Kick an user from a room."
```
Comment (jimmycuadra) [inline src/r0/membership.rs]: a user

File: src/r0/membership.rs [hunk 14]
```diff
@@ -329,8 +425,24 @@ pub mod unban_user {
...
+        fn router_path() -> &'static str {
+            "/_matrix/client/r0/rooms/:room_id/unban"
+        }
+
+        fn name() -> &'static str {
+            "unban_user"
+        }
+
+        fn description() -> &'static str {
+            "unban an user from a room."
```
Comment (jimmycuadra) [inline src/r0/membership.rs]: Unban a

File: src/r0/membership.rs [hunk 15]
```diff
@@ -374,8 +486,24 @@ pub mod ban_user {
...
+        fn router_path() -> &'static str {
+            "/_matrix/client/r0/rooms/:room_id/ban"
+        }
+
+        fn name() -> &'static str {
+            "ban_user"
+        }
+
+        fn description() -> &'static str {
+            "Ban an user from a room."
```
Comment (jimmycuadra) [inline src/r0/membership.rs]: a user

File: src/r0/presence.rs [hunk 16]
```diff
@@ -176,8 +224,25 @@ pub mod get_subscribed_presences {
...
+        fn router_path() -> &'static str {
+            "/_matrix/client/r0/presence/list/:user_id"
+        }
+
+        fn name() -> &'static str {
+            "get_subscribed_presences"
+        }
+
+        fn description() -> &'static str {
+            "get the precence status from the user's subscriptions."
```
Comment (jimmycuadra) [inline src/r0/presence.rs]: capitalize Get

File: src/r0/profile.rs [hunk 17]
```diff
@@ -38,8 +38,24 @@ pub mod get_display_name {
...
+        fn router_path() -> &'static str {
+            "/_matrix/client/r0/profile/:user_id/displayname"
+        }
+
+        fn name() -> &'static str {
+            "get_display_name"
+        }
+
+        fn description() -> &'static str {
+            "Get the display name of an user."
```
Comment (jimmycuadra) [inline src/r0/profile.rs]: a user

File: src/r0/profile.rs [hunk 18]
```diff
@@ -127,8 +159,24 @@ pub mod get_avatar_url {
...
+        fn router_path() -> &'static str {
+            "/_matrix/client/r0/profile/:user_id/avatar_url"
+        }
+
+        fn name() -> &'static str {
+            "get_avatar_url"
+        }
+
+        fn description() -> &'static str {
+            "Get the avatar url name of an user."
```
Comment (jimmycuadra) [inline src/r0/profile.rs]: avatar url name of an user --> avatar URL of a user

File: src/r0/profile.rs [hunk 19]
```diff
@@ -171,8 +219,24 @@ pub mod set_avatar_url {
...
+        fn router_path() -> &'static str {
+            "/_matrix/client/r0/profile/:user_id/avatar_url"
+        }
+
+        fn name() -> &'static str {
+            "set_avatar_url"
+        }
+
+        fn description() -> &'static str {
+            "Set the avatar url of the user."
```
Comment (jimmycuadra) [inline src/r0/profile.rs]: url --> URL

File: src/r0/redact.rs [hunk 20]
```diff
@@ -49,8 +49,24 @@ pub mod redact_event {
...
+        fn router_path() -> &'static str {
+            "/_matrix/client/r0/rooms/:room_id/redact/:event_id/:txn_id"
+        }
+
+        fn name() -> &'static str {
+            "redact_event"
+        }
+
+        fn description() -> &'static str {
+            "Redact an event, stripping all information not critical to the DAG integrity."
```
Comment (jimmycuadra) [inline src/r0/redact.rs]: Probably better to say "event graph" instead of "DAG" here.

File: src/r0/search.rs [hunk 21]
```diff
@@ -222,11 +222,27 @@ pub mod search_events {
...
+        fn router_path() -> &'static str {
+            "/_matrix/client/r0/search"
+        }
+
+        fn name() -> &'static str {
+            "search"
+        }
+
+        fn description() -> &'static str {
+            "Request a server-side search."
```
Comment (jimmycuadra) [inline src/r0/search.rs]: "Search events." might be better

File: src/r0/send.rs [hunk 22]
```diff
@@ -139,8 +171,24 @@ pub mod send_message_event {
...
+        fn router_path() -> &'static str {
+            "/_matrix/client/r0/rooms/:room_id/send/:event_type/:txn_id"
+        }
+
+        fn name() -> &'static str {
+            "send_message_event"
+        }
+
+        fn description() -> &'static str {
+            "Send a message event in a room."
```
Comment (jimmycuadra) [inline src/r0/send.rs]: in --> to

File: src/r0/send.rs [hunk 23]
```diff
@@ -41,14 +41,30 @@ pub mod send_state_event {
...
+        fn router_path() -> &'static str {
+            "/_matrix/client/r0/rooms/:room_id/state/:event_type"
+        }
+
+        fn name() -> &'static str {
+            "send_state_event_for_empty_key"
+        }
+
+        fn description() -> &'static str {
+            "Send a state event in a room associated with the empty state key."
```
Comment (jimmycuadra) [inline src/r0/send.rs]: in --> to

File: src/r0/send.rs [hunk 24]
```diff
@@ -90,8 +106,24 @@ pub mod send_state_event_by_state_key {
...
+        fn router_path() -> &'static str {
+            "/_matrix/client/r0/rooms/:room_id/state/:event_type/:state_key"
+        }
+
+        fn name() -> &'static str {
+            "send_state_event_for_key"
+        }
+
+        fn description() -> &'static str {
+            "Send a state event in a room associated with a given state key."
```
Comment (jimmycuadra) [inline src/r0/send.rs]: in --> to

File: src/r0/session.rs [hunk 25]
```diff
@@ -29,11 +59,27 @@ pub mod login {
...
+        fn router_path() -> &'static str {
+            "/_matrix/client/r0/login"
+        }
+
+        fn name() -> &'static str {
+            "login"
+        }
+
+        fn description() -> &'static str {
+            "User login endpoint."
```
Comment (jimmycuadra) [inline src/r0/session.rs]: "Log in to the homeserver."

File: src/r0/session.rs [hunk 26]
```diff
@@ -55,11 +101,27 @@ pub mod logout {
...
+        fn router_path() -> &'static str {
+            "/_matrix/client/r0/logout"
+        }
+
+        fn name() -> &'static str {
+            "logout"
+        }
+
+        fn description() -> &'static str {
+            "User logout endpoint."
```
Comment (jimmycuadra) [inline src/r0/session.rs]: "Log out of the homeserver."

File: src/r0/sync.rs [hunk 27]
```diff
@@ -156,8 +204,27 @@ pub mod get_member_events {
...
+        fn name() -> &'static str {
+            "get_member_events"
+        }
+
+        fn description() -> &'static str {
+            "Get membership events for a room."
+        }
+
+        fn requires_authentication() -> bool {
+            // not marked as requiring auth in the spec, but
```
Comment (jimmycuadra) [inline src/r0/sync.rs]: Add a TODO here so we can grep for it.

File: src/r0/voip.rs [hunk 28]
```diff
@@ -26,11 +26,27 @@ pub mod get_turn_server_info {
...
+        fn router_path() -> &'static str {
+            "_matrix/client/r0/voip/turnServer"
+        }
+
+        fn name() -> &'static str {
+            "turn_server_info"
+        }
+
+        fn description() -> &'static str {
+            "Credentials for the client to use when initiating VoIP calls."
```
Comment (jimmycuadra) [inline src/r0/voip.rs]: Get credentials

File: src/unversioned.rs [hunk 29]
```diff
@@ -24,11 +24,27 @@ pub mod get_supported_versions {
...
+        fn router_path() -> &'static str {
+            "/_matrix/client/versions"
+        }
+
+        fn name() -> &'static str {
+            "api_version"
+        }
+
+        fn description() -> &'static str {
+            "Versions of the Client API supported by this homeserver."
```
Comment (jimmycuadra) [inline src/unversioned.rs]: Get the versions of the client-server API

File: src/r0/account.rs (hunk 3 above)
Comment (vberger) [inline src/r0/account.rs]: In my mind it was "the currently logged-in account", this endpoint does not allow to deactivate _any_ account, only the one associated with the access token in use

File: src/r0/account.rs (hunk 4 above)
Comment (vberger) [inline src/r0/account.rs]: Same here about "the"

File: src/r0/account.rs (hunk 3 above)
Comment (jimmycuadra) [inline src/r0/account.rs]: Maybe "Deactivate the current user's account"?

Comment (vberger) [inline src/r0/account.rs]: Sounds good! 👍

**Comment by vberger** [comment]
ALright, should be good now 😁

**Comment by jimmycuadra** [comment]
Squash commits and then I'l merge it!

**Pull request merged**